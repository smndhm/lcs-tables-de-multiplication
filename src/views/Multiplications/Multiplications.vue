<template>
  <main>
    <div
      v-if="timer"
      role="timer"
      :style="{
        'animation-duration': `${time}s`,
        'animation-direction': displayResult ? 'reverse' : 'normal',
      }"
    >
      <span class="sr-only">{{ timer }}</span>
    </div>
    <div
      class="operation"
      :class="{ 'operation--display-result': displayResult }"
      v-if="multiplication"
    >
      <span aria-live="polite">
        {{ multiplication.table }}Ã—{{ multiplication.factor }}</span
      ><span
        class="result"
        aria-live="polite"
        :aria-label="
          $t('multiplications.equal', { result: multiplication.result })
        "
        :style="{ visibility: displayResult ? 'visible' : 'hidden' }"
        >={{ multiplication.result }}</span
      >
    </div>
  </main>
</template>

<script>
export default {
  name: "Multiplications",
  data() {
    return {
      factors: [2, 3, 4, 5, 6, 7, 8, 9],
      multiplication: null,
      multiplications: [],
      displayResult: false,
      times: {
        answer: 6,
        result: 3,
      },
      time: null,
      timer: 0,
      interval: null,
    };
  },
  methods: {
    waitFor(seconds) {
      this.time = seconds;
      this.timer = seconds;
      return new Promise((resolve) => {
        this.interval = setInterval(() => {
          this.timer = this.timer - 1;
          if (this.timer === 0) {
            clearInterval(this.interval);
            resolve();
          }
        }, 1000);
        document.addEventListener(
          "mouseup",
          () => {
            this.timer = 0;
            clearInterval(this.interval);
            resolve();
          },
          true
        );
        document.addEventListener(
          "keyup",
          () => {
            this.timer = 0;
            clearInterval(this.interval);
            resolve();
          },
          true
        );
      });
    },
    async start() {
      for (const multiplication of this.multiplications) {
        // display multiplication
        this.multiplication = multiplication;
        // wait for user to answer
        await this.waitFor(this.times.answer);
        // display result
        this.displayResult = true;
        // wait for next multiplication
        await this.waitFor(this.times.result);
        // hide result
        this.displayResult = false;
      }
      // end
      this.$router.push({
        name: "Home",
      });
    },
  },
  mounted() {
    const tables = Array.isArray(this.$route.query.tables)
      ? this.$route.query.tables
      : [this.$route.query.tables];

    // Generate multiplications list
    tables.forEach((table) => {
      table = parseInt(table, 10);
      this.factors.forEach((factor) => {
        this.multiplications.push({
          table,
          factor,
          result: table * factor,
        });
      });
    });

    // shuffle multiplications list
    for (let i = this.multiplications.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.multiplications[i], this.multiplications[j]] = [
        this.multiplications[j],
        this.multiplications[i],
      ];
    }

    // start
    this.start();
  },
};
</script>

<style lang="scss" scoped>
.operation {
  display: flex;
  justify-content: center;
  font-size: 24vw;
  user-select: none;
  font-weight: bold;
}

.result {
  opacity: 0;
  transition: flex 250ms ease-out, opacity 0s;
  width: auto;
  flex: 0;
  overflow: hidden;
}

.operation--display-result .result {
  flex: 1;
  opacity: 1;
  transition: flex 250ms ease-out, opacity 250ms ease-out 250ms;
}

[role="timer"] {
  position: absolute;
  top: 0;
  left: 0;
  height: 0;
  width: 0;
  border-bottom: 0.3rem solid;

  animation-name: timerProgress;
  animation-timing-function: linear;
}

.afficher-resultat [role="timer"] {
  animation-direction: reverse;
}
@keyframes timerProgress {
  from {
    width: 0%;
  }

  to {
    width: 100%;
  }
}
</style>
