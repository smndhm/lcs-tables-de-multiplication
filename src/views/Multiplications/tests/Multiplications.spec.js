import { config, mount } from "@vue/test-utils";
const { axe, toHaveNoViolations } = require("jest-axe");

expect.extend(toHaveNoViolations);

import Multiplications from "../Multiplications";

config.global.mocks = {
  $t: (key) => key,
  $route: { query: { tables: "7" } },
};

describe("Multiplications.vue", () => {
  test("Should renders view", async () => {
    const wrapper = mount(Multiplications, {
      // attachedTo: document.body,
      data() {
        return {
          factors: [2, 3],
        };
      },
    });
    const mockedMultiplication = { table: 8, factor: 7, result: 56 };
    wrapper.vm.multiplication = mockedMultiplication;

    await wrapper.vm.$nextTick();
    let html = wrapper.html();
    expect(html).toMatchSnapshot();
    expect(await axe(html)).toHaveNoViolations();

    document.dispatchEvent(new Event("keyup"));
    await wrapper.vm.$nextTick();
    await wrapper.vm.$nextTick();
    html = wrapper.html();
    expect(html).toMatchSnapshot();
    expect(await axe(html)).toHaveNoViolations();

    await new Promise((r) => setTimeout(r, 3100));
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.multiplication).not.toEqual(mockedMultiplication);

    document.dispatchEvent(new Event("mouseup"));
    await wrapper.vm.$nextTick();

    document.dispatchEvent(new Event("mouseup"));
    await wrapper.vm.$nextTick();
  });

  test("Should manage when query tables il not an array", async () => {
    config.global.mocks.$route.query.tables = ["7"];
    const wrapper = mount(Multiplications);
    expect(wrapper).toBeDefined();
  });
});
