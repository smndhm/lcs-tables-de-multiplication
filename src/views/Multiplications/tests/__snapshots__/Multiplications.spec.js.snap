// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multiplications.vue Should renders view 1`] = `
<main>
  <div role="timer" style="animation-duration: 6s; animation-direction: normal;"><span class="sr-only">6</span></div>
  <div class="operation">8×7<span class="result">=56</span></div>
</main>
`;

exports[`Multiplications.vue Should renders view 2`] = `
<main>
  <div role="timer" style="animation-duration: 3s; animation-direction: reverse;"><span class="sr-only">3</span></div>
  <div class="operation operation--display-result">8×7<span class="result">=56</span></div>
</main>
`;
